crates/sui-types/src/crypto.rs

```diff
@@ -48,6 +48,114 @@ pub type AccountPublicKey = Ed25519PublicKey;
 pub type AccountPrivateKey = Ed25519PrivateKey;
 pub type AccountSignature = Ed25519Signature;
 
+#[enum_dispatch]
+#[derive(Debug)]
+pub enum KeyPair {
+    Ed25519KeyPair,
+    Secp256k1KeyPair,
+}
+
+#[derive(Debug)]
+pub enum PublicKey<'a> {
+    Ed25519KeyPair(&'a Ed25519PublicKey),
+    Secp256k1KeyPair(&'a Secp256k1PublicKey),
+}
+
+impl KeyPair {
+    pub fn public(&self) -> PublicKey<'_> {
+        match self {
+            KeyPair::Ed25519KeyPair(kp) => PublicKey::Ed25519KeyPair(kp.public()),
+            KeyPair::Secp256k1KeyPair(kp) => PublicKey::Secp256k1KeyPair(kp.public()),
+        }
+    }
+}
+
+impl signature::Signer<Signature> for KeyPair {
+    fn try_sign(&self, msg: &[u8]) -> Result<Signature, signature::Error> {
+        match self {
+            KeyPair::Ed25519KeyPair(kp) => kp.try_sign(msg),
```

Comment:
Is there any way to avoid this code duplication for each new scheme we add?

Reply:
due to the associated type not compatible with enum_dispatch, this is the pattern that works, indeed it is quite a few boilerplate - cc @punwai if you have any ideas